### DISCUSSION TITLE
Display colour and resolution issue with esp32-s3-touch-lcd-7

### DISCUSSION THREAD

**Discussion by alexf13e** [slint-ui/slint#6341] [Q&A] [answered]
I've recently purchased a [waveshare esp32-s3-touch-lcd-7](https://www.waveshare.com/wiki/ESP32-S3-Touch-LCD-7) which I hope to use for a home project. I have managed to get a development environment set up with vscode and esp-idf, and have successfully built and flashed programs to the device.

When trying to follow [this guide](https://releases.slint.dev/1.8.0/docs/cpp/mcu/esp_idf) I could not find a support board for my specific device, but found the [component for the touch screen](https://components.espressif.com/components/espressif/esp_lcd_touch_gt911/versions/1.1.1~1?language=en) which also included files for the display. I have managed to get the example program working with all the bsp things replaced and configured manually, based on an example included [in this zip from the waveshare site](https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-7/ESP32_S3_Display_libraries.zip) (ESP-IDF-5.3.0/lvgl_Porting), which was also pre-loaded on the device.

My issue is that the slint test program looks like the picture below, and I haven't been able to figure out a fix for it. The touchscreen does seem to be configured correctly, and the test button responds to being tapped in the correct location, but flickers and turns green in response to being held.
![20240926_183455](https://github.com/user-attachments/assets/ebcc813d-7951-4a55-986e-eff29dc443c5)

The lvgl_porting example program does not have this issue, as shown in the below image. But slint seems to be so much nicer to work with after making a small prototype of my project on desktop, so I would still like to try and make it work on the device.
![20240926_183533](https://github.com/user-attachments/assets/fb593398-61db-48c2-a817-1f2ed438516f)

I have checked my `idf.py menuconfig` and made the settings mentioned on the waveshare page match. I uploaded the relevant files [here](https://github.com/alexf13e/slint-esp-colour-issue) since they're too much to try and include in this post.

My only guess based on how the issue looks is that the pixel data is the wrong size or not aligned properly somehow. The waveshare page says the screen supports 65k colours, which I would assume means they are 16 bit and likely the 565 format used in the slint example (and no other buffer types seem supported by the SlintPlatformConfiguration struct). However, [in the display datasheet](https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-4.3/ST7262.pdf#page=53) there are seemingly references to the pixels having 8 bits per channel such as on the page linked and also page 43. Setting the panel config to 24 bits per pixel doesn't resolve the issue, presumably because the framebuffer is still 16 bits.

Thanks for any help, I really hope I can use slint for the project because it's been so nice to work with so far.

**Comment by tronical** [comment, marked as answer]
Hi! Mostly likely you're running into this issue:

https://releases.slint.dev/1.8.0/docs/cpp/mcu/esp-idf/troubleshoot#wrong-colors-shown

Try setting `color_swap` to false,
here:

https://github.com/alexf13e/slint-esp-colour-issue/blob/ecee2e076aeada6fca3efb92486a41722e7063fd/main/slint-hello-world.cpp#L217

**Comment by alexf13e** [comment]
Thank you, that does appear to have fixed it. I had tried that earlier when I was having a different display issue (not having set psram properly) and didn't think to try it again...
![20240926_215037](https://github.com/user-attachments/assets/9baa3498-cbb5-4148-afc9-f5d1ab2f8bd6)
